pallets/asset-tx-payment/src/tests.rs

```diff
@@ -0,0 +1,819 @@
+			},
+		}.assimilate_storage(&mut t).unwrap();
+		t.into()
+	}
+}
+
+/// create a transaction info struct from weight. Handy to avoid building the whole struct.
+pub fn info_from_weight(w: Weight) -> DispatchInfo {
+	// pays_fee: Pays::Yes -- class: DispatchClass::Normal
+	DispatchInfo { weight: w, ..Default::default() }
+}
+
+fn post_info_from_weight(w: Weight) -> PostDispatchInfo {
+	PostDispatchInfo {
+		actual_weight: Some(w),
+		pays_fee: Default::default(),
+	}
+}
+
+fn post_info_from_pays(p: Pays) -> PostDispatchInfo {
+	PostDispatchInfo {
+		actual_weight: None,
+		pays_fee: p,
+	}
+}
+
+fn default_post_info() -> PostDispatchInfo {
+	PostDispatchInfo {
+		actual_weight: None,
+		pays_fee: Default::default(),
+	}
+}
+
+#[test]
+fn transaction_payment_in_native_possible() {
+	ExtBuilder::default()
+		.balance_factor(10)
+		.base_weight(5)
+		.build()
+		.execute_with(||
+	{
+		let len = 10;
+		let pre = ChargeAssetTxPayment::<Runtime>::from(0, None)
+			.pre_dispatch(&1, CALL, &info_from_weight(5), len)
+			.unwrap();
+		assert_eq!(Balances::free_balance(1), 100 - 5 - 5 - 10);
+
+		assert_ok!(
+			ChargeAssetTxPayment::<Runtime>
+				::post_dispatch(pre, &info_from_weight(5), &default_post_info(), len, &Ok(()))
+		);
+		assert_eq!(Balances::free_balance(1), 100 - 5 - 5 - 10);
+
+		let pre = ChargeAssetTxPayment::<Runtime>::from(5 /* tipped */, None)
+			.pre_dispatch(&2, CALL, &info_from_weight(100), len)
+			.unwrap();
+		assert_eq!(Balances::free_balance(2), 200 - 5 - 10 - 100 - 5);
+
+		assert_ok!(
+			ChargeAssetTxPayment::<Runtime>
+				::post_dispatch(pre, &info_from_weight(100), &post_info_from_weight(50), len, &Ok(()))
+		);
+		assert_eq!(Balances::free_balance(2), 200 - 5 - 10 - 50 - 5);
+	});
+}
+
+#[test]
+fn transaction_payment_in_asset_possible() {
+	let base_weight = 5;
+	ExtBuilder::default()
+		.balance_factor(10)
+		.base_weight(base_weight)
+		.build()
+		.execute_with(||
+	{
+		// work around visibility limitations in the assets pallet
+		let force_create = |asset_id, owner, is_sufficient, min_balance| {
+			Call::Assets(AssetsCall::force_create(asset_id, owner, is_sufficient, min_balance))
+				.dispatch(Origin::root())
+		};
+
+		let len = 10;
+		let asset_id = 1;
+		let owner = 42;
+		let min_balance = 2;
+		let caller = 1;
+		let beneficiary = <Runtime as system::Config>::Lookup::unlookup(caller);
+		assert_ok!(force_create(asset_id, owner, true /* is_sufficient */, min_balance));
+		let balance = 100;
+		assert_ok!(Assets::mint_into(asset_id, &beneficiary, balance));
+		assert_eq!(Assets::balance(asset_id, caller), balance);
+		let weight = 5;
+		// we have a ratio of 1::2 for native vs asset minimum balance
+		let fee = (base_weight + weight + len as u64) * 2;
+		let pre = ChargeAssetTxPayment::<Runtime>::from(0, Some(asset_id))
+			.pre_dispatch(&caller, CALL, &info_from_weight(weight), len)
+			.unwrap();
+		// assert that native balance is not used
+		assert_eq!(Balances::free_balance(caller), balance);
```

Comment:
would be cool not only to assert this, but also assert that the user has zero balance to start with (maybe in another test)

Reply:
:heavy_check_mark: